File: java/src/main/java/org/knguyenc/datastructures/heaps/AbstractHeap.java
package org.knguyenc.datastructures.heaps;

import java.util.Iterator;
import java.util.NoSuchElementException;

public abstract class AbstractHeap<T extends Comparable<T>> implements Iterable<T> {

    private final int DEFAULT_CAPACITY = 50;
    protected final T[] data;
    protected int size;

    @SuppressWarnings("unchecked")
    protected AbstractHeap() {
        data = (T[]) new Comparable[DEFAULT_CAPACITY];
        size = 0;
    }

    @SuppressWarnings("unchecked")
    protected AbstractHeap(int capacity) {
        if (capacity <= 0) capacity = DEFAULT_CAPACITY;
        data = (T[]) new Comparable[capacity];
        size = 0;
    }

    public abstract void buildHeap();

    public abstract void heapify(int index);

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == data.length;
    }

    public void swap(int i, int j) {
        T temp = data[i];
        data[i] = data[j];
        data[j] = temp;
    }

    public abstract void insert(T value);

    public T remove(int index) {
        if (isEmpty()) throw new Error("Heap underflow, can not remove");
        if (index < 0 && index >= size) throw new Error("Invalid index");
        T removed = data[index];
        data[index] = data[--size];
        data[size] = null;
        heapify(index);
        return removed;
    }

    protected static int parent(int index) {
        return (index - 1) / 2;
    }

    protected static int left(int index) {
        return 2 * index + 1;
    }

    protected static int right(int index) {
        return 2 * index + 2;
    }

    public T[] data() {
        return data;
    }

    public int size() {
        return size;
    }

    public void setSize(int size) {
        this.size = size;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private int index = 0;

            @Override
            public boolean hasNext() {
                return index < size;
            }

            @Override
            public T next() {
                if (!hasNext()) throw new NoSuchElementException("No more element in heap");
                return data[index++];
            }
        };
    }

    @Override
    public String toString() {
        int breakLine = 0;
        StringBuilder out = new StringBuilder();
        for (int i = 0; i < size / 2; i++) {
            if (i == 0) out.append("(").append(data[i]).append(" [").append(i).append("])\n");
            if (left(i) < size) out.append("(").append(data[left(i)]).append(" [").append(left(i)).append("])");
            if (right(i) < size) out.append("(").append(data[right(i)]).append(" [").append(right(i)).append("])");
            if (i == breakLine) {
                out.append("\n");
                breakLine = 2 * breakLine + 2;
            } else out.append("  ");
        }
        return out.toString();
    }
}

File: data_structure/heap/abstract_heap.py
"""
Heap
"""
from abc import ABC, abstractmethod
from typing import TypeVar, Generic, Generator, Any

T = TypeVar("T")


class AbstractHeap(ABC, Generic[T]):
    def __init__(self, cap: int = 100):
        self._data: list[T] = [None] * cap
        self._cap = cap
        self._size = 0

    def __iter__(self) -> Generator[T, Any, None]:
        if self._size == 1:
            yield self._data[0]
        elif not self.is_empty():
            for i in range(self._size // 2):
                if i == 0:
                    yield self._data[i]
                if 0 < self._left(i) < self.size:
                    yield self._data[self._left(i)]
                if 0 < self._right(i) < self.size:
                    yield self._data[self._right(i)]

    def __repr__(self) -> str:
        break_line, ret = 0, ""
        for i, v in enumerate(self):
            ret += f"[{v}]({i})"
            if i == break_line:
                ret += "\n"
                break_line = 2 * break_line + 2
            else:
                ret += " "
        return ret

    @property
    def size(self) -> int:
        return self._size

    @size.setter
    def size(self, size: int):
        self._size = size

    def build_heap(self):
        for i in range(self._size // 2, -1, -1):
            self.heapify(i)

    @abstractmethod
    def heapify(self, index: int):
        pass

    @abstractmethod
    def insert(self, value: T):
        pass

    def remove(self, index: int) -> T:
        if self.is_empty():
            raise RuntimeError("Heap underflow")

        if index >= self._size:
            raise RuntimeError("Invalid index")

        removed = self._data[index]
        self._size -= 1
        self._data[index] = self._data[self._size]
        self._data[self._size] = None

        self.heapify(index)
        return removed

    def is_empty(self) -> bool:
        return self._size == 0

    def is_full(self) -> bool:
        return self._size == self._cap

    def swap(self, i: int, j: int):
        self._data[i], self._data[j] = self._data[j], self._data[i]

    @staticmethod
    def _parent(index: int) -> int:
        return (index - 1) // 2

    @staticmethod
    def _left(index: int) -> int:
        return 2 * index + 1

    @staticmethod
    def _right(index: int) -> int:
        return 2 * index + 2
